deltacat/compute/compactor_v2/utils/compaction_utils.py

```diff
@@ -0,0 +1,713 @@
+                    num_hash_groups=params.hash_group_count,
+                    object_store=params.object_store,
+                ),
+                write_to_partition=compacted_partition,
+                compacted_file_content_type=params.compacted_file_content_type,
+                primary_keys=params.primary_keys,
+                sort_keys=params.sort_keys,
+                merge_task_index=index,
+                drop_duplicates=params.drop_duplicates,
+                max_records_per_output_file=params.records_per_compacted_file,
+                enable_profiler=params.enable_profiler,
+                metrics_config=params.metrics_config,
+                s3_table_writer_kwargs=params.s3_table_writer_kwargs,
+                read_kwargs_provider=params.read_kwargs_provider,
+                round_completion_info=round_completion_info,
+                object_store=params.object_store,
+                deltacat_storage=params.deltacat_storage,
+                deltacat_storage_kwargs=params.deltacat_storage_kwargs,
+                delete_strategy=delete_strategy,
+                delete_file_envelopes=delete_file_envelopes,
+                memory_logs_enabled=params.memory_logs_enabled,
+                disable_copy_by_reference=params.disable_copy_by_reference,
+            )
+        }
+
+    merge_tasks_pending = invoke_parallel(
+        items=all_hash_group_idx_to_obj_id.items(),
+        ray_task=mg.merge,
+        max_parallelism=task_max_parallelism,
+        options_provider=merge_options_provider,
+        kwargs_provider=merge_input_provider,
+    )
+    merge_invoke_end = time.monotonic()
+    logger.info(f"Getting {len(merge_tasks_pending)} merge results...")
+    merge_results: List[MergeResult] = ray.get(merge_tasks_pending)
+
+    return merge_results, merge_invoke_end
+
+
+def _process_merge_results(
+    params: CompactPartitionParams, merge_results, mutable_compaction_audit
+) -> tuple[Delta, list[MaterializeResult], dict]:
+    mat_results = []
+    for merge_result in merge_results:
+        mat_results.extend(merge_result.materialize_results)
+
+    mat_results: List[MaterializeResult] = sorted(
+        mat_results, key=lambda m: m.task_index
+    )
+
+    hb_id_to_entry_indices_range = {}
+    file_index = 0
+    previous_task_index = -1
+
+    for mat_result in mat_results:
+        assert (
+            mat_result.pyarrow_write_result.files >= 1
+        ), "Atleast one file must be materialized"
+        assert (
+            mat_result.task_index != previous_task_index
+        ), f"Multiple materialize results found for a hash bucket: {mat_result.task_index}"
+
+        hb_id_to_entry_indices_range[str(mat_result.task_index)] = (
+            file_index,
+            file_index + mat_result.pyarrow_write_result.files,
+        )
+
+        file_index += mat_result.pyarrow_write_result.files
+        previous_task_index = mat_result.task_index
+
+    s3_utils.upload(
```

Comment:
Seems this s3 upload happens in both `_process_merge_results` and `_get_compaction_result`?
Is that intended?